File: src/domain/contracts.ts
export type Capability = "conversation" | "knowledge" | "action";

export interface InboundMessage {
  sessionId: string;
  actorId: string;
  channel: string;
  text: string;
  accountId?: string;
  contactName?: string;
  rawPayload: unknown;
  receivedAt: string;
}

export interface RouteDecision {
  capability: Capability;
  intent: string;
  confidence: number;
  needsKnowledge: boolean;
  statePatch: Partial<ShortTermState>;
  reason: string;
}

export interface CapabilityResult {
  responseText: string;
  statePatch?: Partial<ShortTermState>;
  handoffRequired: boolean;
  artifacts: Record<string, unknown>;
  memoryHints: string[];
}

export interface TurnOutcome {
  capability: Capability;
  intent: string;
  confidence: number;
  responseText: string;
  handoffRequired: boolean;
  stateSnapshot: ShortTermState;
  artifacts: Record<string, unknown>;
}

export interface TurnRecord {
  role: "user" | "assistant";
  text: string;
  timestamp: string;
}

export interface ShortTermState {
  summary: string;
  recentTurns: TurnRecord[];
  activeGoal?: string;
  stage?: string;
  pendingAction?: string;
  lastCapability?: Capability;
  lastIntent?: string;
  continuitySignals: string[];
  turnCount: number;
}

export interface MemoryHit {
  id: string;
  memory: string;
  score: number;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface AddMemoryResult {
  stored: boolean;
  count: number;
}

export interface KnowledgeDocument {
  id: string;
  content: string;
  score: number;
  metadata: Record<string, unknown>;
}

export interface PromptMemorySelection {
  rawRecall: MemoryHit[];
  promptDigest: string;
}

export interface ExecutionContext {
  inbound: InboundMessage;
  shortTermState: ShortTermState;
  memorySelection: PromptMemorySelection;
  knowledge: KnowledgeDocument[];
  routeDecision: RouteDecision;
  traceId: string;
}

export interface RouteTraceDataset {
  traceId: string;
  capability: Capability;
  intent: string;
  confidence: number;
  reason: string;
}

export interface ReplyTraceDataset {
  traceId: string;
  capability: Capability;
  inputText: string;
  responseText: string;
}

File: dspy_service/contracts.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


Capability = Literal["conversation", "knowledge", "action"]
DspyTarget = Literal["route_decision", "conversation"]


class FlexibleModel(BaseModel):
    model_config = ConfigDict(extra="allow")


class InboundMessage(FlexibleModel):
    sessionId: str = ""
    actorId: str = ""
    channel: str = "generic"
    text: str = ""
    correlationId: str | None = None
    parentRunId: str | None = None
    trigger: str | None = None
    accountId: str | None = None
    contactName: str | None = None
    rawPayload: Any = None
    receivedAt: str = ""


class ShortTermState(FlexibleModel):
    summary: str = ""
    recentTurns: list[dict[str, Any]] = Field(default_factory=list)
    activeGoal: str | None = None
    stage: str | None = None
    pendingAction: str | None = None
    lastCapability: Capability | None = None
    lastIntent: str | None = None
    continuitySignals: list[str] = Field(default_factory=list)
    turnCount: int = 0


class PromptMemorySelection(FlexibleModel):
    rawRecall: list[dict[str, Any]] = Field(default_factory=list)
    promptDigest: str = ""


class KnowledgeDocument(FlexibleModel):
    id: str = ""
    content: str = ""
    score: float = 0.0
    metadata: dict[str, Any] = Field(default_factory=dict)


class RouteDecision(FlexibleModel):
    capability: Capability = "conversation"
    intent: str = "general_conversation"
    confidence: float = 0.0
    needsKnowledge: bool = False
    statePatch: dict[str, Any] = Field(default_factory=dict)
    reason: str = ""


class CapabilityResult(FlexibleModel):
    responseText: str = ""
    statePatch: dict[str, Any] | None = None
    handoffRequired: bool = False
    artifacts: dict[str, Any] = Field(default_factory=dict)
    memoryHints: list[str] = Field(default_factory=list)


class RouteDecisionPayload(FlexibleModel):
    inbound: InboundMessage = Field(default_factory=InboundMessage)
    state: ShortTermState = Field(default_factory=ShortTermState)
    promptDigest: str = ""


class ExecutionContext(FlexibleModel):
    inbound: InboundMessage = Field(default_factory=InboundMessage)
    shortTermState: ShortTermState = Field(default_factory=ShortTermState)
    memorySelection: PromptMemorySelection = Field(default_factory=PromptMemorySelection)
    knowledge: list[KnowledgeDocument] = Field(default_factory=list)
    routeDecision: RouteDecision = Field(default_factory=RouteDecision)
    traceId: str = ""


class OptimizeRequest(FlexibleModel):
    target: DspyTarget
    trainsetPath: str | None = None
    valsetPath: str | None = None
    trainset: list[dict[str, Any]] | None = None
    valset: list[dict[str, Any]] | None = None
    auto: Literal["light", "medium", "heavy"] = "light"
    maxBootstrappedDemos: int = 2
    maxLabeledDemos: int = 2
    numThreads: int = 4
    seed: int = 9
    save: bool = True

